`__cleanAttrs`: Clean packages that don't expose all internals (#217243)
NixOS/nixpkgs: doc/stdenv/stdenv.chapter.md

### Special variables {#special-variables}

#### `__cleanAttrs` {#var-__cleanAttrs}

> By default, `mkDerivation` will expose its arguments in the returned package attribute set. This is unnecessary and leads to some confusion and doubt.

Reviewer: ```suggestion
By default, `mkDerivation` will expose its arguments in the returned package attribute set.
This is often unnecessary and may lead to confusion about what's relevant for consumers.
```

We should soften that a bit, because I don't see evidence for the originally definitive-sounding statement.

New packages may pass `__cleanAttrs = true;` to `mkDerivation`, so that it will return a minimal set of package attributes, which package authors can extend via [`passthru`](#var-stdenv-passthru).

Existing packages may be modified to pass `__cleanAttrs = "warn";`, so that the legacy attributes remain available, but they will print a helpful warning when they are accessed. Doing this is best avoided until the packaging function used supports recursively defined arguments like [`mkDerivation` does](#mkderivation-recursive-attributes).

Benefits of `__cleanAttrs`:

 - Users don't have to sift through unnecessary attributes when exploring a package in the `nix repl`.
 - Users can confidently use non-standard attributes that a package provides.
 - Package authors know which variables are intended for use by the builder. They can change the builder environment with confidence that they don't break consumers of their package.
 - It is a little bit more efficient in terms of CPU cycles and memory use.

##### What to do when warned {#warning-package-attr-impl-detail}

> If you encounter the warning `The attribute ... of package ... is an implementation detail`, you are invited to help us explicitly support ways in which a package attribute set may used, so that Nixpkgs contributors and users will be aware of your use case.

Reviewer: This could be phrased to be more immediately actionable.

```suggestion
If you encounter the warning `The attribute ... of package ... is an implementation detail`, please make Nixpkgs contributors and users aware of your particular use case, and help providing explicit ways to use package attribute sets, by opening an [issue](https://github.com/NixOS/nixpkgs/issues) or (https://github.com/NixOS/nixpkgs/pulls).
```
